### ISSUE TITLE
About A-Softmax

### ISSUE THREAD

**Issue by BeginnerW** [wy1iu/LargeMargin_Softmax_Loss#15]
@wy1iu Hello!
The A-Softmax you proposed is also an excellent work. But i encountered a big problem when i used it to fine-tune a face recognition CNN model.  The pre-trained model have already achieved about 98% accuracy on LFW. I have to mention that the 98% model was obtained by training on A-Softmax with SINGLE type. However, when i changed it to QUADRUPLE type to do fine-tuning, after 1w iterations with small learning rate like 0.0002 and 64 batch size, the caffemodel i got was total a mess(like about 50% on LFW). What might be the problem? I wish you could provide me some guidance! Thanks a lot!

**Comment by MDR-EX1000** [comment]
@BeginnerW 
May I ask how do you implement A-Softmax with L-Softmax?  I'm still dealing with L-Softmax to get more experience in tuning the relevant hyper-parameters...

**Comment by BeginnerW** [comment]
@MDR-EX1000 
I just modified relevant files coming from L-Softmax to implement A-Softmax. The A-Softmax layer in prototxt file likes below
###################################
layer {
  name: "ip2"
  type: "LargeMarginInnerProduct"
  bottom: "fc1"
  bottom: "label"
  bottom: "Normalize-W"
  top: "ip2"
  top: "lambda"
#  param {
#    name: "ip2"
#    lr_mult: 1
#  }
  largemargin_inner_product_param {
    num_output: 10565
#    type: SINGLE
#    base: 0
    type: DOUBLE 
    base: 0
    gamma: 0.000025
    power: 35
    iteration: 0
    lambda_min: 0
  }
  include {
    phase: TRAIN
  }
}
May I ask what accuracy do you get on LFW with L-Softmax? Could you share some experience about tuning the hyper-parameters in L-Softmax? I guess my A-Softmax training will benefit from your experience. Thanks!

**Comment by MDR-EX1000** [comment]
@BeginnerW 
My Baseline is 98.68% , up to now I fail to get better result using L-softmax. I think the setting of hyper-parameters can be very tricky ... The loss at convergence is much higher than original softmax...
Some tricks to help convergence:　fine-tune from a base model instead of training from scratch, printing the original softmax loss during train to help adjust learning strategy, set lambda_min when loss goes high, let lambda decrease more smoothly....

**Comment by MDR-EX1000** [comment]
@BeginnerW 
So the only modification is to normalize W? The reported result on MegaFace is really impressive...
It seems to work in another paper! https://arxiv.org/abs/1704.06369

**Comment by BeginnerW** [comment]
@MDR-EX1000 
As I understand it, the modifications are normalizing weights and getting rid of bias term. NormFace is also nice work. Did you try it? What CNN structure do you use and how about your training data? Sorry about the continuous questions.

**Comment by BrettLL** [comment]
@BeginnerW have you ever modified the large-magin-inner-product-layer?? The original layer's bottom num is 2, but you set it 3.
layer {
name: "ip2"
type: "LargeMarginInnerProduct"
bottom: "fc1"
bottom: "label"
bottom: "Normalize-W"
top: "ip2"
top: "lambda"
}

**Comment by MDR-EX1000** [comment]
@BeginnerW 
Nevermind... I haven't tried NormFace but I think it is easier to reproduce.
 The training set is CASIA-WEBFACE and the model structure is a 16-layer CNN like lighn-CNN. The difference in performance can be attribute to the face cropping tools...

**Comment by liangsirrun** [comment]
@MDR-EX1000  you mean you just use Webface to train lightcnn,the accuracy achieve   98.68%? could you share your train_test.prototxt and solver?

**Comment by MDR-EX1000** [comment]
@liangsirrun  Like I said the difference in performance can be attribute to the face cropping tools... The structure is similar to Light CNN B

**Comment by BeginnerW** [comment]
@BrettLL 
Yes, some alterations are made to implement A-Softmax. How about your L-Softmax training?

**Comment by ctgushiwei** [comment]
@MDR-EX1000 could you share your solver to train Light CNN B?

**Comment by BrettLL** [comment]
@BeginnerW I trained the L-softmax almost same performance as softmax when I finetune.

**Comment by wy1iu** [comment]
If you can not make it converge by adjusting the lambda, please use PReLU instead. It will be much easier to converge. I personally use PReLU.

**Comment by wy1iu** [comment]
ReLU sometimes is harder to converge.

**Comment by ctgushiwei** [comment]
@BeginnerW @BrettLL @wy1iu @MDR-EX1000 I trained a the model  use Webface，when the feature dim is 256 or 512 ,the accuracy achieve 97.7%,but when use the 2048 or 4096,the accuracy drop to 95.6%?can you help me to solve this problem?

**Comment by bkj** [comment]
In A-Softmax, is there a reason why, you normalize `W` but not `x`?  Seems like if you really care about the angles, it might be a good idea.  Any thoughts?

**Comment by MDR-EX1000** [comment]
@bkj for a single sample the norm of x doesn't influence the classification result, I think it simply influences the magnitude of gradient rather than the update direction...  But actually, in the sphereFace paper the feature is visualized on a hyper-sphere, it do makes more sense ...

**Comment by NothingL** [comment]
It's very hard to make A-softmax converge when the output classify number is very huge. Such as 300k unique person to train face recognition. Already replace relu with prelu and setting a big lamdamin(30), what else can I do to make it converge?

**Comment by BeginnerW** [comment]
@NothingL What else methods have you tried? Relax other regularization requirements may be work.

**Comment by NothingL** [comment]
@BeginnerW How did you implement A-softmax, just modify w to w/||w|| before each iter's forward? or calculate  ∂E/∂||W|| * ∂||W||/∂W ?

**Comment by BeginnerW** [comment]
I already mentioned that above. I suggest you do the toy example as paper does if you are not sure whether you implement A-softmax correctly.

**Comment by jay2002** [comment]
@BeginnerW Have you trained mnist with a-softmax successfully? In my experiment, the margin is not large enough
